tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,310 @@
+
+                        return Ok(());
+                    } else {
+                        unreachable!(
+                            "acquired the permit after OnceCell value was already initialized"
+                        );
+                    }
+                }
+                _ => {
+                    if !self.initialized() {
+                        panic!(
+                            "couldn't acquire a permit even though OnceCell value is uninitialized"
+                        );
+                    }
+                }
+            }
+        }
+
+        Err(AlreadyInitializedError)
+    }
+
+    /// Tries to initialize the value of the OnceCell using the async function `f`.
+    /// If the value of the OnceCell had already been initialized prior to this call
+    /// a reference to that initialized value is returned
+    ///
+    /// This will deadlock if `f` tries to initialize the cell itself.
+    pub async fn get_or_init<F, Fut>(&self, f: F) -> &T
+    where
+        F: FnOnce() -> Fut,
+        Fut: Future<Output = T>,
+    {
+        if self.initialized() {
+            // SAFETY: once the value is initialized, no mutable references are given out, so
+            // we can give out arbitrarily many immutable references
+            return unsafe { self.get_unchecked() };
+        } else {
+            // After acquire().await we have either acquired a permit while self.value
+            // is still uninitialized, or current thread is awoken after another thread
+            // has intialized the value and closed the semaphore, in which case self.initialized
+            // is true and we don't set the value here
+            match self.semaphore.acquire().await {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        let value = f().await;
+
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!("acquired semaphore after value was already initialized");
+                    }
+                }
+                Err(_) => {
+                    if self.initialized() {
+                        // SAFETY: once the value is initialized, no mutable references are given out, so
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!(
+                            "semaphore acquire call failed when value not already intialized"
```

Comment:
This is a bit unclear.
```suggestion
                            "Semaphore closed, but the OnceCell has not been initialized."
```